src/common/quota/quota.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quota
+
+import (
+	"time"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/common/dao"
+	"github.com/goharbor/harbor/src/common/models"
+)
+
+// Resource ...
+type Resource map[string]int64
+
+// Manager manager for quota
+type Manager struct {
+	reference   string
+	referenceID string
+}
+
+func (m *Manager) getQuota(o orm.Ormer) (*models.Quota, error) {
+	quota := models.Quota{Reference: m.reference, ReferenceID: m.referenceID}
+	err := o.ReadForUpdate(&quota, "reference", "reference_id")
+	if err != nil {
+		return nil, err
+	}
+
+	return &quota, nil
+}
+
+func (m *Manager) getUsage(o orm.Ormer) (*models.QuotaUsage, error) {
+	usage := models.QuotaUsage{Reference: m.reference, ReferenceID: m.referenceID}
+	err := o.ReadForUpdate(&usage, "reference", "reference_id")
+	if err != nil {
+		return nil, err
+	}
+
+	return &usage, nil
+}
+
+func (m *Manager) updateUsage(o orm.Ormer, resource Resource, op operate) error {
+	quota, err := m.getQuota(o)
+	if err != nil {
+		return err
+	}
+
+	usage, err := m.getUsage(o)
+	if err != nil {
+		return err
+	}
+
+	used, err := calcUsed(quota, usage, resource, op)
+	if err != nil {
+		return err
```

Comment:
Should return a specific error for upper layers to determine it's internal server error or just because the request doesn't meet the quota requirement?

Reply:
unsafe error return when request resources failed